### PR TITLE
WIP: Initial work on more scalable repos

### PR DESCRIPTION
**Pull request by alexlarsson**
I've recently been working on various ways to make ostree repos, and in particular the size of the summary file scale better. Almost every operation in flatpak needs an up-to-date version of the remote summary file, and they are both large and update often.

Main reasons they are large are:
 * Inefficient formats
 * Unnecessarily list *all* deltas in summary
 * Many apps make for many refs, especially since the number of refs get multiplied by the number of supported arches.

Also, the more apps we have the more likely one updates triggering a new summary.

My recent work in ostree (https://github.com/ostreedev/ostree/pull/2188) now allows us to completely handle the summary format inside flatpak, which lets us create a new more efficient format, including using deltas to do incremental summary updates.

This is the beginning of that work.

### CODE REVIEW

**Comment by cgwalters** [comment]
I'd be curious to see a high level sketch of the new summary format; I glanced at the code but it wasn't obvious.

**Comment by alexlarsson** [comment]
@cgwalters Its really quite similar to the old one, and I'll add a more complete doc to it later. However, i'll give a brief summary here.

So, we have the old `repo/summary[.sig]` files as before. However, they are now generated by flatpak instead of ostree. Main difference is that you can filter the set of refs in it by arch (goal here is to add arches to flathub but not bloat the compat summary with the new arches).  We also don't add the commit timestamp to the per-ref summary, because flatpak doesn't use it and its quite large.

Secondly we create the new version summaries. These are still still summaries in the sense that ostree could use them as-is (we tell the difference by the `xa.summary-version` metadata key). However they they differ in a few ways:

There are never any deltas. We rely on delta indexes for that.

The way flatpak encodes extra information for each ref is different. Rather than having separate `ref -> data` maps as extra metadata in the topelevel `a{sv}` (i.e `xa.cache`and `xa.sparse-cache` we put the same data in the actual per-ref metadata. This allows us to not repeat the ref strings all the time, but we couldn't do this before due to how the ostree summary generation API works.

We split the summary into subsets. Mainly by arch, but the goal is to also be able to create additional subsets (i.e. all gnu-free apps on flathub). This makes each individual sub-summary smaller, and each client is in practice only going to need one of them. To index the separate summaries we have a single summary index which references the subsets by a string name to a digest. 

There are some global metadata in the summary, like the repo title, icon, way to change the url and the gpg keys, etc. These move from the individual summary metadata dict to the index dict as we need only store these once. Also we move the global summary timestamp from the subsummary metadata to the dict so that the subsummaries are "reproducible" over time.

Each individual subsummary is not signed, instead we sign the index.

In terms of repo layout it goes like:

```
repo/summary.idx
repo/summary.idx.sig
repo/summaries/$SHA256
repo/summaries/$SHA256-$SAH256.delta
```

Where the `summary.idx` is signed and contains a map from "subset name" to information about each subsummary plus a standard metadata dict. The per-subsummary information currently contains the digest of the subsummary and a history of previous digests for that subset. The history is used on the server side to know which deltas to generate and can be used by the client as a hint of whether its worth looking for a summary delta.

The deltas are a very simple format. It contains a set of chunks, each either of "Reuse N bytes from old file", "Skip N bytes from old files", "Use N bytes of supplied data". Then all of this is gzip:ed. Due to the sorted format of the summary it is very easy and cheap to create such diffs, and the code for applying them is very simple and easy to review.

**Comment by alexlarsson** [comment]
I merged this because it seems to be in a good base state now. Will continue documenting it and other work separately.

**Review by debarshiray — COMMENTED** [review]
Coverity flagged some problems in this code.  I am still reading the code and playing with it, to understand what the original intent was, but meanwhile ...

File: common/flatpak-utils.c [hunk 1]
```diff
@@ -1377,2179 +1459,3677 @@ flatpak_decompose_ref (const char *full_ref,
...
-    cache_version = 0;
+  new_len = var_ref_map_get_length (new_refs);
+  old_len = var_ref_map_get_length (old_refs);
 
-  if (cache_version < FLATPAK_XA_CACHE_VERSION)
-    return; /* We need to rebuild the cache with the current version */
+  new_i = old_i = 0;
+  while (new_i < new_len && old_i < old_len)
+    {
+      if (new_i == new_len)
```
[the remark this answers is not in the document]
Comment (debarshiray) [reply common/flatpak-utils.c:5024]: I removed this branch in https://github.com/flatpak/flatpak/pull/4932

File: common/flatpak-utils.c [hunk 2]
```diff
@@ -1377,2179 +1459,3677 @@ flatpak_decompose_ref (const char *full_ref,
...
+        {
+          /* Just old left */
+          old_entry = var_ref_map_get_at (old_refs, old_i);
+          old_ref = var_ref_map_entry_get_ref (old_entry);
+          old_i++;
+          diff_consume_block (&data,
+                              -1, 0,
+                              (const guchar *)new_entry.base - (const guchar *)new.base, new_entry.size);
+        }
+      else if (old_i == old_len)
```
[the remark this answers is not in the document]
Comment (debarshiray) [reply common/flatpak-utils.c:5034]: Removed in https://github.com/flatpak/flatpak/pull/4932

**Pull request merged**